### ISSUE TITLE
Failed to import Cucumber execution: AxiosError: Request failed with status code 400

### ISSUE THREAD

**Issue by caique-rd** [Qytera-Gmbh/cypress-xray-plugin#151]
Hello @csvtuda how are you doing? Sorry to bother again with another problem, but I found some problems using cucumber to send the result to Xray.

Right now my settings are:
```ts
// cypress.config.ts
import { defineConfig } from "cypress";
import { addCucumberPreprocessorPlugin } from "@badeball/cypress-cucumber-preprocessor";
const createBundler = require("@bahmutov/cypress-esbuild-preprocessor");
const createEsbuildPlugin = require("@badeball/cypress-cucumber-preprocessor/esbuild").createEsbuildPlugin;
import { addXrayResultUpload, configureXrayPlugin, syncFeatureFile } from "cypress-xray-plugin";
import fix from "cypress-on-fix"
export default defineConfig({
e2e: {
async setupNodeEvents(on, config) {
const fixedOn = fix(on);
    await configureXrayPlugin(config, {
        plugin: {
          debug: true,
        },
        cucumber: {
          featureFileExtension: ".feature",
          uploadFeatures: true
        },
        jira: {
          projectKey: "RPF",
          url: "https://my-project.atlassian.net",
          testExecutionIssueKey: process.env.CYPRESS_TEST_EXECUTION_KEY,
        },
        xray: {
          uploadResults: process.env.CYPRESS_TEST_EXECUTION_KEY !== '',
        },
      });

      await addCucumberPreprocessorPlugin(fixedOn, config);
        await addXrayResultUpload(fixedOn);
        fixedOn("file:preprocessor", async (file) => {
            await syncFeatureFile(file);
            const cucumberPlugin = createBundler({
                plugins: [createEsbuildPlugin(config)],
            });
            return cucumberPlugin(file);
        });

      return config;   
    },
    specPattern: 'cypress/e2e/features/*.feature',
}}
})
```

And I'm getting the following error:
```
│ Cypress Xray Plugin │ INFO    │ Importing execution (Cucumber)...
│ Cypress Xray Plugin │ DEBUG   │ Request:  /services/ui/logs/1692659980332_POST_https_xray.cloud.getxray.app_api_v2_import_execution_cucumber_multipart_request.json
│ Cypress Xray Plugin │ INFO    │ Waiting for https://xray.cloud.getxray.app/api/v2 to respond... (10 seconds)
│ Cypress Xray Plugin │ INFO    │ Waiting for https://xray.cloud.getxray.app/api/v2 to respond... (20 seconds)
│ Cypress Xray Plugin │ DEBUG   │ Response: /services/ui/logs/1692660010176_POST_https_xray.cloud.getxray.app_api_v2_import_execution_cucumber_multipart_response.json
| Cypress Xray Plugin │ ERROR   │ Failed to import Cucumber execution: AxiosError: Request failed with status code 400
│ Cypress Xray Plugin │ WARNING │ Execution results import was skipped. Skipping remaining tasks
│ Cypress Xray Plugin │ ERROR   │ Complete error logs have been written to: /services/ui/logs/importExecutionCucumberMultipartError.json
```

Looking at the response from `https_xray.cloud.getxray.app_api_v2_import_execution_cucumber_multipart_response.json` I get:
```json
{
   "message":"Request failed with status code 400",
   "name":"AxiosError",
   "stack":"AxiosError: Request failed with status code 400\n    at settle (/services/ui/node_modules/cypress-xray-plugin/node_modules/axios/lib/core/settle.js:19:12)\n    at IncomingMessage.handleStreamEnd (/services/ui/node_modules/cypress-xray-plugin/node_modules/axios/lib/adapters/http.js:570:11)\n    at IncomingMessage.emit (node:events:525:35)\n    at IncomingMessage.emit (node:domain:489:12)\n    at endReadableNT (node:internal/streams/readable:1359:12)\n    at processTicksAndRejections (node:internal/process/task_queues:82:21)"
}
```

**Comment by csvtuda** [comment]
Hey @caique-raidiam,

are there additional information in `/services/ui/logs/importExecutionCucumberMultipartError.json` besides the error message you already posted? Sometimes the logs contain messages like `projectKey is missing`, which could be quite useful.

Are you perhaps able to create a minimal Cucumber test so I can reproduce the problem? It hopefully does not necessarily need to target your real application, generic steps and assertions like `expect(true).to.be.true`) should usually be enough.

**Comment by caique-raidiam** [comment]
Hey @csvtuda thanks for your return in `/services/ui/logs/importExecutionCucumberMultipartError.json` I have:
```
{"error":{"message":"Request failed with status code 400","name":"AxiosError","stack":"AxiosError: Request failed with status code 400\n    at settle (/services/ui/node_modules/cypress-xray-plugin/node_modules/axios/lib/core/settle.js:19:12)\n    at IncomingMessage.handleStreamEnd (/services/ui/node_modules/cypress-xray-plugin/node_modules/axios/lib/adapters/http.js:570:11)\n    at IncomingMessage.emit (node:events:525:35)\n    at IncomingMessage.emit (node:domain:489:12)\n    at endReadableNT (node:internal/streams/readable:1359:12)\n    at processTicksAndRejections (node:internal/process/task_queues:82:21)"
```
And
```
"method":"post","url":"https://xray.cloud.getxray.app/api/v2/import/execution/cucumber/multipart","data":{"_overheadLength":320,"_valueLength":461393,"_valuesToMeasure":[],"writable":false,"readable":true,"dataSize":0,"maxDataSize":2097152,"pauseStreams":true,"_released":true,"_streams":[],"_currentStream":null,"_insideLoop":false,"_pendingNext":false,"_boundary":"--------------------------256205210432716330847642","_events":{},"_eventsCount":3}},"code":"ERR_BAD_REQUEST","status":400},"response":{"error":"There are no valid tests imported"}}
```

I'll try to generate a minimal Cucumber test for you and I'll come back later

**Comment by csvtuda** [comment]
Thanks for the quick response. The plugin doesn't seem to upload the Cucumber test results (maybe it can't find the JSON report generated by the Cucumber preprocessor?) Would you mind sharing your your [Cucumber preprocessor configuration](https://github.com/badeball/cypress-cucumber-preprocessor/blob/4120c63277bbc64d349e2bb518b32b6270da66a0/docs/configuration.md)?

**Comment by caique-raidiam** [comment]
Good catch @csvtuda maybe this is the problem, I'm saving the `cucumber.json` file in the path `cypress/generate-reports`  because this file was being saved outside the Cypress folder in another directory, in the UI directory, and I don't want it to be inside the UI folder, but even removing this output and using it as the second option, I still have the same problem
```
"cypress-cucumber-preprocessor": {
    "e2e": {
      "stepDefintions": "cypress/support/step_definitions/[filepath].{js,mjs,ts,tsx}"
    },
    "json": {
      "enabled": true,
      "output": "cypress/generate-reports/cucumber-report.json"
    }
  },
```
second option:
```
"cypress-cucumber-preprocessor": {
    "e2e": {
      "stepDefintions": "cypress/support/step_definitions/[filepath].{js,mjs,ts,tsx}"
    },
    "json": {
      "enabled": true,
      "output": "cypress/cucumber-report.json"
    }
  }
```

**Comment by csvtuda** [comment]
I think your configuration is fine. Can you double check if you've tagged all your scenarios and scenario outlines in your feature files correctly? There _must_ exist a test issue for every tag (I was able to reproduce your error by tagging scenarios with non-existent issues).


For example, if there exists a test issue `CYP-123`, but the scenario is tagged with `CYP-132` by accident and `CYP-132` does not exist, your error will occur.

**Comment by caique-raidiam** [comment]
Hey @csvtuda thanks for your help, but I still get the error, I've created a example here https://github.com/caique-raidiam/cypress-xray-plugin

So in this repository, I just have one test and this RPF-1678 exists in my Xray/Jira Project as a test case. The plugin is finding the test issue in Jira and is updating this issue like the print above, but I still get the same error after the test finish

```
Feature: Active sessions

    @TestName:RPF-1678
    Scenario: RPF-1678 Checking the existing columns in active sessions page
        When I navigate to the active sessions page
        Then The system should show me the following columns
        | ID |
        | User Email |
        | Issued At |
        | Expires At |
        | Delete |
```
![Captura de Tela 2023-08-22 às 18 55 51](https://github.com/Qytera-Gmbh/cypress-xray-plugin/assets/139126939/f3ca654c-e4f8-4f5f-be0c-a5d8d99803de)

![Captura de Tela 2023-08-22 às 18 58 33](https://github.com/Qytera-Gmbh/cypress-xray-plugin/assets/139126939/6e8b731a-b1e0-43b3-920e-b95984abef61)

**Comment by csvtuda** [comment]
Great, thank you for the detailed example! I am able to reproduce the issue and will look into it.

**Comment by caique-raidiam** [comment]
Thank you so much for all your support

**Comment by csvtuda** [comment]
Alright I cannot reproduce it, but we can try to skip the plugin as a middleman. I hope you've got `curl` installed for sending requests to the Xray API manually.

Can you try running the following shell script? You can save it inside your `cypress-xray-plugin` directory as a file called `import.sh`:

```sh
# Acquire bearer token.
# Don't forget to replace <xray-client-id> and <xray-client-secret> with your actual credentials.
TOKEN=$(curl --silent --location 'https://xray.cloud.getxray.app/api/v2/authenticate' \
--header 'Content-Type: application/json' \
--data '{
    "client_id":     "xray-client-id",
    "client_secret": "xray-client-secret"
}')
# Remove quotes from token.
TOKEN="${TOKEN//\"/}"
# Send import request.
curl --silent --location 'https://xray.cloud.getxray.app/api/v2/import/execution/cucumber/multipart' \
--header "Authorization: Bearer $TOKEN" \
--form 'results=@"cypress/generate-reports/cucumber-report.json"' \
--form 'info=@"cypress/generate-reports/info.json"'
```

You can run it using:

```sh
sh import.sh
```

The second request `curl` sends (the actual import request) requires two JSON files, one containing the Cucumber report and one containing information about the execution issue to create.

The report JSON is simply the one from your repository (copy-pasted here):
```json
// cypress/generate-reports/cucumber-report.json
[
  {
    "description": "",
    "elements": [
      {
        "description": "",
        "id": "active-sessions;rpf-1678-checking-the-existing-columns-in-active-sessions-page",
        "keyword": "Scenario",
        "line": 4,
        "name": "RPF-1678 Checking the existing columns in active sessions page",
        "steps": [
          {
            "arguments": [],
            "keyword": "When ",
            "line": 5,
            "name": "I navigate to the active sessions page",
            "match": {
              "location": "not available:0"
            },
            "result": {
              "status": "passed",
              "duration": 14000000
            }
          },
          {
            "arguments": [
              {
                "rows": [
                  {
                    "cells": [
                      "ID"
                    ]
                  },
                  {
                    "cells": [
                      "User Email"
                    ]
                  },
                  {
                    "cells": [
                      "Issued At"
                    ]
                  },
                  {
                    "cells": [
                      "Expires At"
                    ]
                  },
                  {
                    "cells": [
                      "Delete"
                    ]
                  }
                ]
              }
            ],
            "keyword": "Then ",
            "line": 6,
            "name": "The system should show me the following columns",
            "match": {
              "location": "not available:0"
            },
            "result": {
              "status": "passed",
              "duration": 10000000
            }
          }
        ],
        "tags": [
          {
            "name": "@TestName:RPF-1678",
            "line": 3
          }
        ],
        "type": "scenario"
      }
    ],
    "id": "active-sessions",
    "line": 1,
    "keyword": "Feature",
    "name": "Active sessions",
    "tags": [],
    "uri": "cypress/e2e/features/active-sessions.feature"
  }
]
```

For the info JSON, you can use the following:

```json
// cypress/generate-reports/info.json
{
    "fields": {
      "project": {
        "key": "RPF"
      },
      "summary": "Debug Execution Issue",
      "description": "For debugging https://github.com/Qytera-Gmbh/cypress-xray-plugin/issues/151",
      "issuetype": {
        "name": "Test Execution"
      }
    },
    "xrayFields": {}
  }
```

Executing the shell script _should_ result in the following output:

```json
{"id":"12345","key":"RPF-12345","self":"https://<domain>.atlassian.net/rest/api/2/issue/12345"}
```

`RPF-12345` should then be the created test execution issue.

Can you confirm if the error still occurs?

**Comment by caique-raidiam** [comment]
Hey @csvtuda I'll try this today and send here later the result, thanks for you help

**Comment by caique-raidiam** [comment]
Hey @csvtuda with this script I get the following error
```
{"error":"There are no valid tests imported"}
```

**Comment by csvtuda** [comment]
Hey @caique-raidiam,

thank you so much for trying it out. This at least tells me that it's not a plugin issue. Something in your results JSON is not correct. I would like to tell you to double check that the issue actually exists, but you've attached the screenshot already.

The only way I can reliably reproduce your error is:

- by setting the tag in the results JSON to the tag of a nonexistent issue, in your case maybe:

  ```json
  "tags": [
    {
      "name": "@TestName:RPF-123456",
      "line": 3
    }
  ]
  ```
- by setting the tag in the results JSON to the tag of an issue that is _not an Xray test issue_ (but yours looks like it is one)

I don't know what else to do here. The only thing I'd suggest to do is [contact Xray support](https://jira.getxray.app/servicedesk/customer/portal/2), honestly (they've always helped me).

Just tell them you're calling the `/api/v2/import/execution/cucumber/multipart` endpoint and attach both JSONs so that they can retrace what's going on. Maybe it's worth a shot.

**Comment by caique-rd** [comment]
Hey @csvtuda thank you so much for you help, I'll continue my investigation today and I'll ask Xray support for help, I'll let you know if I found anything

**Comment by SebastienCorrigan** [comment]
Hello @csvtuda

I am having the same issue as @caique-rd, although I have a manual work around that may give a clue. In summary, I believe the issue is in the @TestName tag. 

Even though the docs say to use `@TestName:issuekey ` for Jira cloud instances, the upload only works when the report uses `@issuekey`. I wonder if certain versions of Jira cloud only accept `@issuekey`.

Here's the test I did:

Pre conditions:
✅ Issue PRO-100 exists in Jira (cloud instance)

1. Run a minimal cypress cucumber test

`
Feature: Minimal Feature

    @TestName:PRO-100
    Scenario: Minimal Scenario
    Given minimal Given
    When minimal When
    Then minimal Then
`
`
Given("minimal Given", () => {})
When("minimal When", () => {})
Then("minimal Then", () => {
    expect(false).to.be.true
})
`

2. The test fails with the error of this issue:  `Failed to import Cucumber execution: AxiosError: Request failed with status code 400`

```bash
│ Cypress Xray Plugin │ INFO    │ Jira username and API token found. Setting up Jira cloud basic auth credentials
│ Cypress Xray Plugin │ INFO    │ Xray client ID and client secret found. Setting up Xray cloud JWT credentials
│ Cypress Xray Plugin │ INFO    │ Authenticating to: https://xray.cloud.getxray.app/api/v2/authenticate...

│ Cypress Xray Plugin │ INFO    │ Fetching necessary Jira issue type information in preparation for Cucumber result uploads...

│ Cypress Xray Plugin │ INFO    │ Preprocessing feature file cypress/e2e/energy/features/minimal.feature...

│ Cypress Xray Plugin │ INFO    │ Uploading Cucumber test results...
│ Cypress Xray Plugin │ ERROR   │ Failed to import Cucumber execution: AxiosError: Request failed with status code 400
│ Cypress Xray Plugin │ WARNING │ Execution results import was skipped. Skipping remaining tasks
│ Cypress Xray Plugin │ ERROR   │ Complete error logs have been written to [PATH]/logs/importExecutionCucumberMultipartError.json
```

3. Edit the cucumber-report.json, remove 'TestName' in `tags[0].name`

before edit:
```json
        "tags": [
          {
            "name": "@TestName:PRO-100",
            "line": 3
          }
        ],
`

after edit: 
```
        "tags": [
          {
            "name": "@PRO-100",
            "line": 3
          }
        ],
```

4. Use the Xray REST api `https://xray.cloud.getxray.app/api/v2/import/execution/cucumber` to upload the cucumber-report 
5. Import is successful


Note that if I remove `TestName:` from the tag in the `.feature` file, I get an error that indicates that it is required

```

│ Cypress Xray Plugin │ INFO    │ Uploading Cucumber test results...
│ Cypress Xray Plugin │ WARNING │ Skipping result upload for scenario: Minimal Scenario
│ Cypress Xray Plugin │ WARNING │   
│ Cypress Xray Plugin │ WARNING │   No test issue keys found in tags of scenario: Minimal Scenario
│ Cypress Xray Plugin │ WARNING │   You can target existing test issues by adding a corresponding tag:
│ Cypress Xray Plugin │ WARNING │   
│ Cypress Xray Plugin │ WARNING │   @TestName:PRO-123
│ Cypress Xray Plugin │ WARNING │   Scenario: Minimal Scenario
│ Cypress Xray Plugin │ WARNING │     Before undefined
```

**Comment by csvtuda** [comment]
Hey @caique-rd,

It's been some time, but with the help of @SebastienCorrigan we were able to find a solution (see #275 and [here](../blob/main/CHANGELOG.md#600)). The fix is available in version `6.0.0` and above.

Basically, what you need to do is configure the Cucumber prefixes your Xray instance uses to link scenarios and backgrounds to existing Jira issues:

```ts
async setupNodeEvents(on, config) {
  await configureXrayPlugin(
    config,
    {
      cucumber: {
        prefixes: {
          precondition: "PRECONDITION_", // placeholder, use what's configured in you Xray instance
          test: "TEST_"                  // placeholder, use what's configured in you Xray instance
        }
      }
    }
  );
}
```

You can read up on it in the [documentation](https://qytera-gmbh.github.io/projects/cypress-xray-plugin/section/configuration/cucumber/#prefixes).
